drivers/sht2x/sht2x.c

```diff
@@ -0,0 +1,476 @@
+
+    /* Acquire exclusive access */
+    i2c_acquire(BUS);
+
+    DEBUG("[SHT2x] write command: addr=%02x cmd=%02x\n", ADDR, (uint8_t)command);
+    i2c_result = i2c_write_byte(BUS, ADDR, (uint8_t)command, 0);
+
+    i2c_release(BUS);
+
+    if (i2c_result != 0) {
+        return SHT2X_ERR_I2C;
+    }
+
+    return SHT2X_OK;
+}
+
+/*
+ * Returns temperature in centi DegC.
+ */
+int16_t sht2x_read_temperature(const sht2x_t* dev)
+{
+    uint16_t raw_value;
+    int i2c_result;
+    if (dev->params.measure_mode == SHT2X_MEASURE_MODE_NO_HOLD) {
+        i2c_result = read_sensor_poll(dev, eTempNoHoldCmd, &raw_value);
+    } else {
+        i2c_result = read_sensor(dev, eTempHoldCmd, &raw_value);
+    }
+    if (i2c_result != SHT2X_OK) {
+        return INT16_MIN;
+    }
+    return (-46.85 + 175.72 / 65536.0 * raw_value) * 100;
+}
+
+/*
+ * Returns humidity in centi %RH (i.e. the percentage times 100).
+ */
+uint16_t sht2x_read_humidity(const sht2x_t *dev)
+{
+    uint16_t raw_value;
+    int i2c_result;
+    if (dev->params.measure_mode == SHT2X_MEASURE_MODE_NO_HOLD) {
+        i2c_result = read_sensor_poll(dev, eRHumidityNoHoldCmd, &raw_value);
+    } else {
+        i2c_result = read_sensor(dev, eRHumidityHoldCmd, &raw_value);
+    }
+    if (i2c_result != SHT2X_OK) {
+        return 0;
+    }
+    return 100 * (-6.0 + 125.0 / 65536.0 * raw_value);
+}
+
+static int _sht2x_add_ident_byte(uint8_t * buffer, size_t buflen, uint8_t b, size_t ix)
+{
+    if (ix < buflen) {
+        buffer[ix++] = b;
+    }
+    return ix;
+}
+
+int sht2x_read_ident(const sht2x_t *dev, uint8_t * buffer, size_t buflen)
+{
+    uint8_t data1[8];        /* SNB_3, CRC, SNB_2, CRC, SNB_1, CRC, SNB_0, CRC */
+    uint8_t data2[6];        /* SNC_1, SNC_0, CRC, SNA_1, SNA_0, CRC */
+    size_t ix;
+    int res;
+
+    i2c_acquire(BUS);
+    res = i2c_read_regs(BUS, ADDR,
+                        first_mem_addr, data1, sizeof(data1), I2C_REG16);
+    i2c_release(BUS);
+    if (res < 0) {
+        return -1;
+    }
+
+    DEBUG("[SHT2x] ident (1): %02x %02x %02x %02x\n", data1[0], data1[1], data1[2], data1[3]);
+    DEBUG("[SHT2x] ident (1): %02x %02x %02x %02x\n", data1[4], data1[5], data1[6], data1[7]);
+    for (size_t ix = 0; ix < sizeof(data1); ix += 2) {
+        if (sht2x_checkcrc(&data1[ix], 1, data1[ix + 1]) != 0) {
+            DEBUG("[SHT2x] checksum error first (ix=%d)\n", ix);
+        }
+    }
+
+    i2c_acquire(BUS);
+    res = i2c_read_regs(BUS, ADDR,
+                        second_mem_addr, data2, sizeof(data2), I2C_REG16);
+    i2c_release(BUS);
+    if (res < 0) {
+        return -1;
+    }
+    DEBUG("[SHT2x] ident (2): %02x %02x %02x\n", data2[0], data2[1], data2[2]);
+    DEBUG("[SHT2x] ident (2): %02x %02x %02x\n", data2[3], data2[4], data2[5]);
+    for (size_t ix = 0; ix < sizeof(data2); ix += 3) {
+        if (sht2x_checkcrc(&data2[ix], 2, data2[ix + 2]) != 0) {
+            DEBUG("[SHT2x] checksum error, second (ix=%d)\n", ix);
+        }
+    }
```

Comment:
Not so sure about compiler optimization here but would it be reasonable to guard everything from line 208-229 by `#if ENABLE_DEBUG`?

Reply:
Hmm, I want to add a return SHT2X_ERR_CRC. In that case the loop is still needed.

Reply:
Makes sense.